### PR TITLE
refine balance mechanism by pruning code and support urgent stopping

### PR DESCRIPTION
**Pull request by MrPresent-Han**
related: #23454

### CODE REVIEW

**Comment by xiaofan-luan** [comment]
overall it looks good to me except for the data race.

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

File: internal/querycoordv2/checkers/balance_checker.go [hunk 1]
```diff
@@ -18,34 +18,128 @@ package checkers
...
+	stoppingCollections := make([]int64, 0)
+	for _, cid := range loadedCollections {
+		replicas := b.meta.ReplicaManager.GetByCollection(cid)
+	outer:
+		for _, replica := range replicas {
+			for _, nodeID := range replica.GetNodes() {
+				isStopping, _ := b.nodeManager.IsStoppingNode(nodeID)
+				if isStopping {
+					stoppingCollections = append(stoppingCollections, cid)
+					break outer
```
Comment (congqixia) [inline internal/querycoordv2/checkers/balance_checker.go:82]: maybe we shall not break here. By default all loaded collections will be affected by a stopping node if no resource group is used. If we break here, all collection will need multiple round to recover.

File: internal/querycoordv2/checkers/balance_checker.go [hunk 2]
```diff
@@ -18,34 +18,128 @@ package checkers
...
+		}
+	}
+	//do stopping balance only in this round
+	if len(stoppingCollections) > 0 {
+		return stoppingCollections
+	}
+	//no stopping balance and auto balance is disabled, return empty collections for balance
 	if !Params.QueryCoordCfg.AutoBalance {
-		return ret
+		return make([]int64, 0)
```
Comment (congqixia) [inline internal/querycoordv2/checkers/balance_checker.go:93]: NIT: 
```suggestion
		return nil
```

File: internal/querycoordv2/checkers/balance_checker.go [hunk 3]
```diff
@@ -18,34 +18,128 @@ package checkers
...
+	}
+
+	// scheduler is handling segment task, skip
+	if b.scheduler.GetSegmentTaskNum() != 0 {
+		return make([]int64, 0)
```
[the remark this answers is not in the document]
Comment (MrPresent-Han) [reply internal/querycoordv2/checkers/balance_checker.go:98]: changed

File: internal/querycoordv2/checkers/balance_checker.go (hunk 2 above)
Comment (MrPresent-Han) [reply internal/querycoordv2/checkers/balance_checker.go:93]: changed

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**